src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +612,216 @@ public static string GetTempPath(string workingDir)
             return Environment.GetEnvironmentVariable("SYSTEMROOT");
         }
     }
+
+    internal interface IServerMutex : IDisposable
+    {
+        bool TryLock(int timeoutMs);
+        void Unlock();
+        bool IsDisposed { get; }
+    }
+
+    internal sealed class FileMutex : IDisposable
+    {
+        public readonly bool OwnsMutex;
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool IsLocked { get; private set; }
+
+        internal static string GetMutexDirectory()
+        {
+            var homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Path.GetTempPath();
```

Comment:
We should avoid `GetTempPath` directly because it will differ from invocation to invocation of MSBuild. Instead we should prefer `BuildServerConnection.GetTempPath`. It removes the biggest sources of flakiness.